File: CTDash/simulation/game_loop.py
# [BUILD FIRST]
# GameLoop — the main simulation tick.
# Each tick advances all timers by one time step (defined in config.TICK_SECONDS).
# Managers are called in a fixed order each tick — order matters.
#
# Correct tick order:
#   1. shift_timer  — advance clock, check for hourly events
#   2. spawn        — maybe spawn a new patient
#   3. transport    — advance transport state machines
#   4. queue        — advance wait timers
#   5. contrast     — advance contrast timers
#   6. scanner      — advance scan/cooldown timers
#   7. scoring      — check for penalties, update score display

from classes.patient import PatientState
from config import HOLDING_PROPER_SLOTS
from data.acuity_table import ACUITY_TABLE
from simulation.shift_timer import ShiftTimer


class GameLoop:
    def __init__(self, managers: dict, exam_catalog: dict, duration_hours: int = None):
        # managers: dict of {name: manager_instance}
        # e.g. {"spawn": SpawnManager(...), "scanner": ScannerManager(...), ...}
        self.timer        = ShiftTimer(duration_hours) if duration_hours else ShiftTimer()
        self.managers     = managers
        self.exam_catalog = exam_catalog

        # Master patient registry: patient_id -> Patient.
        self._patients = {}
        # Patients that have already been charged the overflow penalty.
        self._overflow_charged = set()

    def run(self) -> int:
        """Main loop — runs until shift is over. Returns final score."""
        while not self.timer.is_shift_over:
            self._tick()
        self._end_shift()
        return self.managers["scoring"].score

    def _tick(self):
        # 1. Advance clock.
        new_hour = self.timer.tick()
        gs = self.timer.game_second
        gh = self.timer.game_hour

        # 2. Spawn — may add patients to queue_manager.
        self.managers["spawn"].tick(gh, gs)

        # Sync newly spawned patients into our master registry and start transport.
        for pid, patient in list(self.managers["queue"]._patients.items()):
            if pid not in self._patients:
                self._patients[pid] = patient
                self.managers["transport"].assign_transport(patient)

        # 3. Transport — advance state machines; sets IN_HOLDING on delivery.
        self.managers["transport"].tick()

        # 4. Queue — increment wait timers for all queued patients.
        self.managers["queue"].tick(gs)

        # 5. Contrast — count down oral contrast timers.
        self.managers["contrast"].tick()

        # 6. Start oral contrast for newly arrived patients that need it.
        self._order_contrast_if_needed()

        # 7. Scanner — advance scan/cooldown timers.
        self.managers["scanner"].tick(self.exam_catalog)

        # 8. Process patients whose all exams are done.
        self._handle_completed_scans()

        # 9. Assign highest-priority ready patients to idle scanners.
        self._try_assign_to_scanners()

        # 10. Apply wait penalties.
        self._apply_wait_penalties()

        # 11. Overflow check.
        self._check_overflow()

        # 12. Staffing — fires only on new game-hour boundary.
        if new_hour:
            self.managers["staffing"].tick(float(gh))

    def _end_shift(self):
        remaining = [
            p for p in self._patients.values()
            if p.state not in (PatientState.COMPLETED, PatientState.REFUSED)
        ]
        final_score = self.managers["scoring"].finalise(remaining)
        print(f"\n=== Shift Over ===")
        print(f"Exams completed : {self.managers['scoring'].exams_completed}")
        print(f"Holdovers       : {self.managers['scoring'].holdovers}")
        print(f"Final score     : {final_score}")

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _order_contrast_if_needed(self):
        """Kick off oral contrast for IN_HOLDING patients whose exam requires it."""
        contrast_mgr = self.managers["contrast"]
        for patient in self._patients.values():
            if patient.state != PatientState.IN_HOLDING:
                continue
            if patient.current_exam_index >= len(patient.exam_list):
                continue
            exam_key = patient.exam_list[patient.current_exam_index]
            exam     = self.exam_catalog[exam_key]
            if exam.oral_contrast:
                contrast_mgr.start_oral_contrast(patient)

    def _try_assign_to_scanners(self):
        """Assign highest-priority ready patients to idle scanners."""
        scanner_mgr  = self.managers["scanner"]
        contrast_mgr = self.managers["contrast"]

        ready = sorted(
            [p for p in self._patients.values()
             if p.state in (PatientState.IN_HOLDING, PatientState.CONTRAST_READY,
                            PatientState.INJECTOR_READY)
             and p.current_exam_index < len(p.exam_list)],
            key=lambda p: (p.acuity, -p.wait_timer),
        )

        for patient in ready:
            exam_key = patient.exam_list[patient.current_exam_index]
            exam     = self.exam_catalog[exam_key]

            # If IV contrast needed and injector not yet filled, fill it first.
            if patient.state == PatientState.CONTRAST_READY and exam.iv_contrast:
                contrast_mgr.fill_injector(patient)
                continue

            assigned = scanner_mgr.try_assign(patient, self.exam_catalog)
            if assigned:
                # Remove from queue once assigned; no longer needs wait-timer tracking.
                self.managers["queue"]._patients.pop(patient.patient_id, None)

    def _handle_completed_scans(self):
        """Handle patients whose entire exam list is finished."""
        scanner_mgr = self.managers["scanner"]
        for patient in scanner_mgr.completed_patients:
            self.managers["scoring"].exam_completed(patient.acuity)
            self.managers["transport"].begin_leaving(patient)
        scanner_mgr.completed_patients.clear()
        # TransportManager.tick() marks LEAVING patients COMPLETED on delivery.

    def _apply_wait_penalties(self):
        """Deduct points for patients waiting past their acuity threshold."""
        for patient in self._patients.values():
            if patient.state in (PatientState.COMPLETED, PatientState.REFUSED,
                                  PatientState.SCANNING, PatientState.COOLDOWN,
                                  PatientState.LEAVING):
                continue
            threshold = ACUITY_TABLE[patient.acuity]["wait_threshold"]
            excess = patient.wait_timer - threshold
            if excess > 0 and excess % 60 == 0:
                self.managers["scoring"].apply_wait_penalty(patient.acuity, 60)

    def _check_overflow(self):
        """One-time flat penalty when a patient first occupies an overflow slot."""
        holding = sorted(
            [p for p in self._patients.values() if p.state == PatientState.IN_HOLDING],
            key=lambda p: (p.acuity, -p.wait_timer),
        )
        for i, patient in enumerate(holding):
            if i >= HOLDING_PROPER_SLOTS:
                if patient.patient_id not in self._overflow_charged:
                    self._overflow_charged.add(patient.patient_id)
                    self.managers["scoring"].apply_overflow_penalty()

File: CTDash_Unity/Assets/Scripts/Simulation/GameLoop.cs
// GameLoop — the main simulation tick.
// Each tick advances all timers by one game-second (GameConfig.TICK_SECONDS).
// Managers are called in a fixed order each tick — order matters.
//
// Tick order:
//   1. ShiftTimer     — advance clock, check for hourly events
//   2. SpawnManager   — maybe spawn a new patient
//   3. TransportManager — advance transport state machines
//   4. QueueManager   — advance wait timers
//   5. ContrastManager — advance contrast timers
//   6. ScannerManager — advance scan/cooldown timers
//   7. Scoring        — check for penalties, update score

using System.Collections.Generic;

namespace CTDash
{
    public class GameLoop
    {
        public readonly ShiftTimer       Timer;
        public readonly SpawnManager     SpawnMgr;
        public readonly QueueManager     QueueMgr;
        public readonly TransportManager TransportMgr;
        public readonly ContrastManager  ContrastMgr;
        public readonly ScannerManager   ScannerMgr;
        public readonly ScoringManager   ScoringMgr;
        public readonly StaffingManager  StaffingMgr;

        // Master patient registry.
        private readonly Dictionary<string, Patient> _patients        = new Dictionary<string, Patient>();
        private readonly HashSet<string>             _overflowCharged = new HashSet<string>();

        private readonly Dictionary<string, Exam> _examCatalog;

        public GameLoop(
            SpawnManager    spawnMgr,
            QueueManager    queueMgr,
            TransportManager transportMgr,
            ContrastManager contrastMgr,
            ScannerManager  scannerMgr,
            ScoringManager  scoringMgr,
            StaffingManager staffingMgr,
            Dictionary<string, Exam> examCatalog,
            int durationHours = GameConfig.GAME_DURATION_HOURS)
        {
            SpawnMgr     = spawnMgr;
            QueueMgr     = queueMgr;
            TransportMgr = transportMgr;
            ContrastMgr  = contrastMgr;
            ScannerMgr   = scannerMgr;
            ScoringMgr   = scoringMgr;
            StaffingMgr  = staffingMgr;
            _examCatalog = examCatalog;
            Timer        = new ShiftTimer(durationHours);
        }

        // Call once per Unity Update (or FixedUpdate) to advance one game-tick.
        public void Tick()
        {
            if (Timer.IsShiftOver) return;

            // 1. Advance clock.
            bool newHour = Timer.Tick();
            int  gs      = Timer.GameSecond;
            int  gh      = Timer.GameHour;

            // 2. Spawn.
            SpawnMgr.Tick(gh, gs);

            // Sync newly spawned patients into master registry and start transport.
            foreach (var kvp in QueueMgr.Patients)
            {
                if (!_patients.ContainsKey(kvp.Key))
                {
                    _patients[kvp.Key] = kvp.Value;
                    TransportMgr.AssignTransport(kvp.Value);
                }
            }

            // 3. Transport.
            TransportMgr.Tick();

            // 4. Queue wait timers.
            QueueMgr.Tick();

            // 5. Contrast timers.
            ContrastMgr.Tick();

            // 6. Order oral contrast for newly arrived patients that need it.
            OrderContrastIfNeeded();

            // 7. Scanner timers.
            ScannerMgr.Tick(_examCatalog);

            // 8. Process patients whose entire exam list is finished.
            HandleCompletedScans();

            // 9. Assign highest-priority ready patients to idle scanners.
            TryAssignToScanners();

            // 10. Wait penalties.
            ApplyWaitPenalties();

            // 11. Overflow check.
            CheckOverflow();

            // 12. Staffing — fires only on new game-hour boundary.
            if (newHour)
                StaffingMgr.Tick((float)gh);
        }

        public int EndShift()
        {
            var remaining = new List<Patient>();
            foreach (var p in _patients.Values)
            {
                if (p.State != PatientState.Completed && p.State != PatientState.Refused)
                    remaining.Add(p);
            }
            return ScoringMgr.Finalise(remaining);
        }

        // --- Helpers -------------------------------------------------------

        private void OrderContrastIfNeeded()
        {
            foreach (var patient in _patients.Values)
            {
                if (patient.State != PatientState.InHolding) continue;
                if (patient.CurrentExamIndex >= patient.ExamList.Count) continue;

                string examKey = patient.ExamList[patient.CurrentExamIndex];
                if (_examCatalog[examKey].OralContrast)
                    ContrastMgr.StartOralContrast(patient);
            }
        }

        private void TryAssignToScanners()
        {
            var ready = new List<Patient>();
            foreach (var p in _patients.Values)
            {
                bool isReady = p.State == PatientState.InHolding
                            || p.State == PatientState.ContrastReady
                            || p.State == PatientState.InjectorReady;

                if (isReady && p.CurrentExamIndex < p.ExamList.Count)
                    ready.Add(p);
            }

            // Highest priority first (lowest acuity number, then longest wait).
            ready.Sort((a, b) =>
            {
                if (a.Acuity != b.Acuity) return a.Acuity.CompareTo(b.Acuity);
                return b.WaitTimer.CompareTo(a.WaitTimer);
            });

            foreach (var patient in ready)
            {
                string examKey = patient.ExamList[patient.CurrentExamIndex];
                var    exam    = _examCatalog[examKey];

                // Fill injector if IV contrast needed and contrast is ready.
                if (patient.State == PatientState.ContrastReady && exam.IvContrast)
                {
                    ContrastMgr.FillInjector(patient);
                    continue;
                }

                bool assigned = ScannerMgr.TryAssign(patient, _examCatalog);
                if (assigned)
                    QueueMgr.Patients.Remove(patient.PatientId);
            }
        }

        private void HandleCompletedScans()
        {
            foreach (var patient in ScannerMgr.CompletedPatients)
            {
                ScoringMgr.ExamCompleted();
                TransportMgr.BeginLeaving(patient);
            }
            ScannerMgr.CompletedPatients.Clear();
        }

        private void ApplyWaitPenalties()
        {
            foreach (var patient in _patients.Values)
            {
                if (patient.State == PatientState.Completed  ||
                    patient.State == PatientState.Refused    ||
                    patient.State == PatientState.Scanning   ||
                    patient.State == PatientState.Cooldown   ||
                    patient.State == PatientState.Leaving) continue;

                int threshold = AcuityData.Table[patient.Acuity].WaitThreshold;
                int excess    = patient.WaitTimer - threshold;
                if (excess > 0 && excess % 60 == 0)
                    ScoringMgr.ApplyWaitPenalty(patient.Acuity, 60);
            }
        }

        private void CheckOverflow()
        {
            var inHolding = new List<Patient>();
            foreach (var p in _patients.Values)
                if (p.State == PatientState.InHolding) inHolding.Add(p);

            inHolding.Sort((a, b) =>
            {
                if (a.Acuity != b.Acuity) return a.Acuity.CompareTo(b.Acuity);
                return b.WaitTimer.CompareTo(a.WaitTimer);
            });

            for (int i = 0; i < inHolding.Count; i++)
            {
                if (i >= GameConfig.HOLDING_PROPER_SLOTS)
                {
                    if (_overflowCharged.Add(inHolding[i].PatientId))
                        ScoringMgr.ApplyOverflowPenalty();
                }
            }
        }
    }
}
